### ISSUE TITLE
Composer phar commands linked against strange Symfony Command?

### ISSUE THREAD

**Issue by exteon** [composer/composer#10117]
```php
    class OnUpdateCommand extends BaseCommand
    {
        public const COMMAND_NAME = 'web3:on-update';

        protected function configure()
        {
            $this
                ->setName(static::COMMAND_NAME)
                ->setHidden(true);
        }

        protected function execute(
            InputInterface $input,
            OutputInterface $output
        ): int {
            return 0;
        }
    }
```

```
composer web3:on-update
PHP Fatal error:  Uncaught Error: Call to undefined method OnUpdateCommand::setHidden()
```

```
bin/composer web3:on-update
```
```
composer self-update --no-plugins
You are already using the latest available Composer version 2.1.8 (stable channel).
```

`$this instanceof \Symfony\Component\Console\Command\Command` seems to hold true for phar `composer`, so maybe it's just a really, really, really out-of-date symfony?

**Comment by stof** [comment]
hidden commands are a feature that exist only in symfony/console 3.2+
The composer phar contains symfony/console 2.8.52 due to the need to support PHP 5.3

**Comment by exteon** [comment]
Interesting... I do wonder though, ever since I started working on composer plugins, why the phar does not defer to the local repository composer as soon as it's installed... it's so counterintuitive that `composer` and `bin/composer` may be inconsistent with each other

**Comment by stof** [comment]
@exteon when using composer, most of the time, you don't install composer itself as a local dependency. If even if you do it, it would be too late in most cases (plugins are allowed to hook early, not only after installation is done, which would limit them a lot).
Making the phar a wrapper around an install from source is a chicken-and-egg problem, as the source of composer relies on composer to install the dependencies

**Comment by exteon** [comment]
I see, kindof the same problem root as https://github.com/composer/composer/issues/10101
Ok, so closing this as it's more of a scheduling issue than a bug.

**Comment by stof** [comment]
@exteon even splitting the before-update and after-update would not solve the issue regarding the symfony/console version. It might solve the issue regarding having 2 different versions of the plugin, but it won't change the code of composer itself (which does not change during an update).
The whole point of composer plugins is to get loaded inside the composer process, not to be installed in a custom build of composer where the lock file is generated based on both composer dependencies and the plugin dependencies and then using that custom build to run composer. This means that plugins have to deal with the version of dependencies locked in the composer phar. The only time such "custom build" setup is usually done is when *developing* plugins, as installing the composer/composer source code helps to have the IDE completion. But that's not producing an equivalent environment and so is not the best setup to *test* plugins. Testing plugins should ideally be done by running the composer phar and loading the plugin.

**Comment by exteon** [comment]
So based on your last comment I backtracked my steps and found that somewhere in my dependencies in a very inadvertent place I had buried a `require composer/composer`. From that point on, I **always** had `bin/composer` present so I assumed that when using plugins, a local composer install is always present and I took this as a design trait. Thanks for your explanation, it really helped me understand now how all this is supposed to work.
